File: pset1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int     main(void)
{
    long    card;
    int     ones = 0;
    int     twos = 0;
    int     click = 0;
    int     id = 0;
    int     digits = 0;
    long    card_copy;
    int     pre_twos;

    do
    {
        card = get_long("Number: ");
    }
    while (card < 0);
    card_copy = card;
    while (card_copy > 0)
    {
        card_copy = card_copy / 10;
        digits++;
    }
    if (digits == 13 || digits == 15 || digits == 16)
    {
        while (card > 0)
        {
            if (card < 100 && !id)
            {
                id = card;
            }
            if (click == 0)
            {
                ones += card % 10;
                card = card / 10;
                click = 1;
            }
            else if (click == 1)
            {
                pre_twos = (card % 10) * 2;
                if (pre_twos > 9)
                {
                    pre_twos = pre_twos % 10 + (pre_twos / 10);
                }
                twos += pre_twos;
                card = card / 10;
                click = 0;
            }
        }
        if ((twos + ones) % 10 == 0)
        {
            if ((id == 34 || id == 37) && digits == 15)
            {
                printf("AMEX\n");
            }
            else if ((id >= 51 && id <= 55) && digits == 16)
            {
                printf("MASTERCARD\n");
            }
            else if ((id >= 40 && id <= 49) && (digits == 13 || digits == 16))
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit.py
from cs50 import get_int
ones = 0
twos = 0
click = 0
idx = 0
digits = 0

# Read number
card = get_int("Number: ")
while (card < 0):
    card = get_int("Number: ")

# Count digits
card_copy = card
while card_copy > 0:
    card_copy = card_copy // 10
    digits += 1

# Check validity
if digits == 13 or digits == 15 or digits == 16:
    # Loop over card
    while card > 0:
        if card < 100 and idx == 0:
            idx = card
        if click == 0:
            ones += card % 10
            card = card // 10
            click = 1
        elif click == 1:
            pre_twos = (card % 10) * 2
            if pre_twos > 9:
                pre_twos = pre_twos % 10 + (pre_twos // 10)
            twos += pre_twos
            card = card // 10
            click = 0
    # Define output
    if (twos + ones) % 10 == 0:
        if (idx == 34 or idx == 37) and digits == 15:
            print("AMEX")
        elif (idx >= 51 and idx <= 55) and digits == 16:
            print("MASTERCARD")
        elif (idx >= 40 and idx <= 49) and (digits == 13 or digits == 16):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")

